src/hotspot/share/runtime/lockStack.inline.hpp

```diff
@@ -61,32 +65,123 @@ inline void LockStack::push(oop o) {
   verify("pre-push");
   assert(oopDesc::is_oop(o), "must be");
   assert(!contains(o), "entries must be unique");
-  assert(can_push(), "must have room");
+  assert(!is_full(), "must have room");
   assert(_base[to_index(_top)] == nullptr, "expect zapped entry");
   _base[to_index(_top)] = o;
   _top += oopSize;
   verify("post-push");
 }
 
-inline void LockStack::remove(oop o) {
+inline oop LockStack::bottom() const {
+  assert(to_index(_top) > 0, "must contain an oop");
+  return _base[0];
+}
+
+inline bool LockStack::is_empty() const {
+  return to_index(_top) == 0;
+}
+
+inline bool LockStack::is_recursive(oop o) const {
+  if (!VM_Version::supports_recursive_lightweight_locking()) {
+    return false;
+  }
+
+  // This will succeed iff there is a consecutive run of oops on the
+  // lock-stack with a length of at least 2.
+
+  assert(contains(o), "entries must exist");
+  int end = to_index(_top);
+  // Start iterating from the top because the runtime code is more
+  // interested in the balanced locking case when the top oop on the
+  // lock-stack matches o. This will cause the for loop to break out
+  // in the first loop iteration if it is non-recursive.
+  for (int i = end - 1; i > 0; i--) {
+    if (_base[i - 1] == o && _base[i] == o) {
+      return true;
+    }
+    if (_base[i] == o) {
+      // o can only occur in one consecutive run on the lock-stack.
```

Comment:
I'm not sure that the claim on L103 is always true. If we have a lock stack like this:

```
_base[end - 1] = o1;
_base[end - 2] = o2;
_base[end - 3] = o1;
_base[end - 4] = o1;
```

When our `o == o1` we don't have a recursive run on the top-most part of
the lock stack, but we do have one that's lower down. L103 isn't correct in
this case, but that doesn't matter because we actually care about whether
the top most run is recursive. I think L103 can be deleted and the rest of
the comment is okay.

Reply:
Thanks for the explanation (and reminder about interleaved recursive locking).